Chat flow bug fixing

alexanishchenko88 opened the pull request:
#### 📲 What

Chat Flow Bug fixing allowing chat skip and updating the answers structure from firebase.

#### 🤔 Why
		
Bug fixing to ensure the correct chat flow
		
#### 🛠 How
		
Updated the data we are storing into firebase after chat configuration finishes. Now instead of simple map we will have Map<String, Map<String, String>>

Added:
ChatResponseKeys.dart

Updated:
All the files related with Map structure update

Deleted:
-

#### ✅ Acceptance criteria

- [ ] Design Review for UI with BA completed. 
- [ ] Manually tested and verified on Android device/emulator (min. Lollipop 5.1)
- [ ] Showcase video / automation test
- [ ] Passing all tests
- [ ] Rebased/merged with latest changes from development and re-tested
- [ ] Removed unsed comments. TODOs must have JIRA for future work.

#### Coding Standards Checklist
- [ ] unit tests 80% coverage on testable code (functions, methods, class)
- [ ] (Architecture) redux state, reducers, actions, middleware defined under lib/redux
- [ ] ui elements defined under lib/ui. 
- [ ] Strings ready for localisation (e.g. defined in lib/utils/strings.dart)
- [ ] Assets (images/icons path) defined in lib/utils/assets.dart
- [ ] Colors defined in lib/utils/colors.dart
- [ ] Dimensions (ie margins padding) defined in lib/utils/dimens.dart
- [ ] TextStyles defined in lib/utils/styles.dart

Recommended Style Guide: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo

#### 🕵️‍♂️ How to test

Notes for QA

#### Reviewers

@farmsmart/amido @farmsmart/wamf

alexanishchenko88 committed:
Chat flow, reset process fixed

What:
Reset process has been fixed. Previously it was doing the back navigation without changing the page state.
Now it changes the state to cancelled and navigates back

Added:
-

Updated:
Changes the state to cancelled and navigates back on tap reset button. Executes _cancel(context)

Deleted:
-

alexanishchenko88 committed:
Chat Flow Bug fixing allowing chat skip and updating the answers structure form firebase.

What:
Updated the data we are storing into firebase after chat configuration finishes. Now instead of simple map we will have Map<String, Map<String, String>>

Added:
ChatResponseKeys.dart

Updated:
All the files related with Map structure update

Deleted:
-

Pending:
Recommendation apparently is not working, should be checked.

alexanishchenko88 commented on lib/chat/utils/ChatResponseKeys.dart:
```diff
@@ -0,0 +1,6 @@
+  static const keyId = "id";
+  static const keyTitle = "title";
+  static const keyValue = "value";
+}
```
As commented right above, not sure about this class, if we should persist that or maybe move it into some other place.

LeeMatthewHiggins replied:
we normally would make it private at the point of use, and call it "_Fields"

LeeMatthewHiggins replied:
we don't need to repeat the context "key"

alexanishchenko88 replied:
Ok, changed, moved to private _Fields class at the point of usage, ChatResponseKeys.dart has been removed, **key** annotation at the beginning of each property removed.

LeeMatthewHiggins replied:
👍

alexanishchenko88 committed:
Fixing recommendations assignment, checking nullability before getting the value from map value which is another map

What:
Check is map value by provided key is not null just before getting the value from this parent value map to avoid null pointer exception.

Added:
-

Updated:
RecommentadionEngine.dart

Deleted:

alexanishchenko88 committed:
PR comments review

What:
ChatResponseKeys.dart has been removed and the content has been replaced to _Fields private class in the point of usage.

Added:
-

Updated:
lib/model/bloc/recommendations/RecommendationEngine.dart
lib/model/bloc/startup/ChatResponseToPlotInfoTransformer.dart
lib/ui/profile/Profile.dart

Deleted:
lib/chat/utils/ChatResponseKeys.dart

LeeMatthewHiggins reviewed (commented):
is this still WIP or ready to be reviewed?

LeeMatthewHiggins committed:
Added trainer option

LeeMatthewHiggins reviewed (approved):
Updated and rebased
